File: ios/VestaQuickAdd/Native/PriceSockets.swift
import Foundation

/// The web app's three real-time price feeds, ported: Binance WS + Gate.io WS
/// for crypto (Gate covers exchange tokens Binance doesn't list — GT, OFC),
/// Alpaca WS (iex) for US stock trades. Connected while the app is foreground,
/// torn down in background; each socket reconnects itself after 5s on failure.
final class PriceSocketCenter: @unchecked Sendable {
    static let alpacaKey = "PK7UABSLFNCFJH2ICVM2YW3A53"
    static let alpacaSecret = "qSy9wj3hBvbBhMJUR3Kb64n7SRn1KfHAg8ktrEZmVWJ"

    /// Guards every stored property below.
    ///
    /// URLSession delivers completion handlers on a CONCURRENT queue, so the
    /// reconnect path mutated `tasks` from several threads at once. That races
    /// Array's copy-on-write realloc and corrupts the buffer — it crashed with
    /// EXC_BAD_ACCESS inside _swift_release_dealloc during an Alpaca reconnect
    /// (which retries every 5s whenever the US market is closed).
    private let lock = NSLock()

    private var active = false
    private var tasks: [URLSessionWebSocketTask] = []
    private var pingTask: Task<Void, Never>?

    /// symbol ("BTCUSDT") → token ("BTC")
    private var binanceMap: [String: String] = [:]
    /// pair ("GT_USDT") → token ("GT")
    private var gateMap: [String: String] = [:]
    private var stockSymbols: [String] = []

    private var onCrypto: (@Sendable (String, Double) -> Void)?
    private var onStock: (@Sendable (String, Double) -> Void)?
    private var tickCounts: [String: Int] = [:]

    private func countTick(_ feed: String, detail: String) {
        lock.lock()
        tickCounts[feed, default: 0] += 1
        let first = tickCounts[feed] == 1
        lock.unlock()
        if first { log("\(feed) first tick: \(detail)") }
    }

    /// DEBUG-only visibility — readable via `simctl launch --console`.
    private func log(_ message: String) {
        #if DEBUG
        print("[sockets] \(message)")
        #endif
    }

    func start(
        binanceMap: [String: String],
        gateMap: [String: String],
        stockSymbols: [String],
        onCrypto: @escaping @Sendable (String, Double) -> Void,
        onStock: @escaping @Sendable (String, Double) -> Void
    ) {
        stop()
        lock.lock()
        active = true
        self.binanceMap = binanceMap
        self.gateMap = gateMap
        self.stockSymbols = stockSymbols
        self.onCrypto = onCrypto
        self.onStock = onStock
        lock.unlock()
        log("start binance=\(binanceMap.count) gate=\(gateMap.count) stocks=\(stockSymbols.count)")
        connectBinance()
        connectGate()
        connectAlpaca()
    }

    func stop() {
        lock.lock()
        active = false
        pingTask?.cancel()
        pingTask = nil
        let open = tasks
        tasks.removeAll()
        lock.unlock()
        for task in open { task.cancel(with: .goingAway, reason: nil) }
    }

    /// Snapshot of `active`, for the reconnect guards.
    private var isActive: Bool {
        lock.lock(); defer { lock.unlock() }
        return active
    }

    // MARK: Plumbing

    private func open(_ url: URL, onOpen: ((URLSessionWebSocketTask) -> Void)? = nil,
                      onMessage: @escaping (URLSessionWebSocketTask, String) -> Void,
                      reconnect: @escaping () -> Void) {
        let task = URLSession.shared.webSocketTask(with: url)
        lock.lock()
        guard active else { lock.unlock(); return } // stopped mid-reconnect
        tasks.append(task)
        lock.unlock()
        task.resume()
        onOpen?(task)
        receiveLoop(task, onMessage: onMessage, reconnect: reconnect)
    }

    private func receiveLoop(_ task: URLSessionWebSocketTask,
                             onMessage: @escaping (URLSessionWebSocketTask, String) -> Void,
                             reconnect: @escaping () -> Void) {
        task.receive { [weak self] result in
            guard let self, self.isActive else { return }
            switch result {
            case .success(let message):
                if case .string(let text) = message { onMessage(task, text) }
                self.receiveLoop(task, onMessage: onMessage, reconnect: reconnect)
            case .failure(let error):
                self.log("socket failed: \(error.localizedDescription) — retrying in 5s")
                // Socket died — retry after a beat if we're still supposed to
                // be live. The old task stays cancelled; a new one replaces it.
                Task { [weak self] in
                    try? await Task.sleep(for: .seconds(5))
                    guard let self, self.isActive else { return }
                    reconnect()
                }
            }
        }
    }

    private func json(_ text: String) -> [String: Any]? {
        text.data(using: .utf8)
            .flatMap { try? JSONSerialization.jsonObject(with: $0) as? [String: Any] }
    }

    // MARK: Binance — combined miniTicker stream

    private func connectBinance() {
        lock.lock(); let binanceMap = self.binanceMap; lock.unlock()
        guard !binanceMap.isEmpty else { return }
        let streams = binanceMap.keys
            .map { "\($0.lowercased())@miniTicker" }
            .sorted()
            .joined(separator: "/")
        guard let url = URL(string: "wss://stream.binance.com:9443/stream?streams=\(streams)")
        else { return }
        open(url, onMessage: { [weak self] _, text in
            guard let self,
                  let root = self.json(text),
                  let data = root["data"] as? [String: Any],
                  let symbol = data["s"] as? String,
                  let token = { self.lock.lock(); defer { self.lock.unlock() }; return self.binanceMap[symbol] }(),
                  let close = (data["c"] as? String).flatMap(Double.init)
            else { return }
            self.countTick("binance", detail: symbol)
            self.onCrypto?(token, close)
        }, reconnect: { [weak self] in self?.connectBinance() })
    }

    // MARK: Gate.io — spot.tickers channel

    private func connectGate() {
        lock.lock(); let gateMap = self.gateMap; lock.unlock()
        guard !gateMap.isEmpty, let url = URL(string: "wss://api.gateio.ws/ws/v4/")
        else { return }
        open(url, onOpen: { [weak self] task in
            guard let self else { return }
            let subscribe: [String: Any] = [
                "time": Int(Date().timeIntervalSince1970),
                "channel": "spot.tickers",
                "event": "subscribe",
                "payload": Array(gateMap.keys),
            ]
            if let data = try? JSONSerialization.data(withJSONObject: subscribe),
               let text = String(data: data, encoding: .utf8) {
                task.send(.string(text)) { _ in }
            }
            // Gate closes idle connections — app-level ping keeps it alive.
            self.lock.lock()
            self.pingTask?.cancel()
            self.pingTask = Task { [weak task] in
                while !Task.isCancelled {
                    try? await Task.sleep(for: .seconds(20))
                    let ping = "{\"time\":\(Int(Date().timeIntervalSince1970)),\"channel\":\"spot.ping\"}"
                    task?.send(.string(ping)) { _ in }
                }
            }
            self.lock.unlock()
        }, onMessage: { [weak self] _, text in
            guard let self,
                  let root = self.json(text),
                  root["channel"] as? String == "spot.tickers",
                  root["event"] as? String == "update",
                  let result = root["result"] as? [String: Any],
                  let pair = result["currency_pair"] as? String,
                  let token = { self.lock.lock(); defer { self.lock.unlock() }; return self.gateMap[pair] }(),
                  let last = (result["last"] as? String).flatMap(Double.init)
            else { return }
            self.countTick("gate", detail: pair)
            self.onCrypto?(token, last)
        }, reconnect: { [weak self] in self?.connectGate() })
    }

    // MARK: Alpaca — iex trades

    private func connectAlpaca() {
        lock.lock(); let stockSymbols = self.stockSymbols; lock.unlock()
        guard !stockSymbols.isEmpty,
              let url = URL(string: "wss://stream.data.alpaca.markets/v2/iex")
        else { return }
        open(url, onOpen: { task in
            let auth = "{\"action\":\"auth\",\"key\":\"\(Self.alpacaKey)\",\"secret\":\"\(Self.alpacaSecret)\"}"
            task.send(.string(auth)) { _ in }
        }, onMessage: { [weak self] task, text in
            guard let self,
                  let data = text.data(using: .utf8),
                  let messages = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
            else { return }
            for message in messages {
                switch message["T"] as? String {
                case "success" where (message["msg"] as? String) == "authenticated":
                    let subscribe: [String: Any] = [
                        "action": "subscribe", "trades": self.stockSymbols,
                    ]
                    if let payload = try? JSONSerialization.data(withJSONObject: subscribe),
                       let text = String(data: payload, encoding: .utf8) {
                        task.send(.string(text)) { _ in }
                    }
                case "t":
                    if let symbol = message["S"] as? String,
                       let price = message["p"] as? Double {
                        self.onStock?(symbol, price)
                    }
                default:
                    break
                }
            }
        }, reconnect: { [weak self] in self?.connectAlpaca() })
    }
}

File: android/app/src/main/java/com/piyawatpm/vesta/data/PriceSockets.kt
package com.piyawatpm.vesta.data

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.TimeUnit

/**
 * The web app's three real-time price feeds, ported: Binance WS + Gate.io WS
 * for crypto (Gate covers exchange tokens Binance doesn't list — GT, OFC),
 * Alpaca WS (iex) for US stock trades. Connected while the app is foreground,
 * torn down in background; each socket reconnects itself after 5s on failure.
 * Mirrors ios PriceSockets.swift.
 */
class PriceSocketCenter {
    companion object {
        // Paper-trading keys — market data only; same convention as the web
        // bundle and the iOS app (see lib/hooks/use-alpaca-ws.ts).
        const val ALPACA_KEY = "PK7UABSLFNCFJH2ICVM2YW3A53"
        const val ALPACA_SECRET = "qSy9wj3hBvbBhMJUR3Kb64n7SRn1KfHAg8ktrEZmVWJ"
    }

    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS) // websockets stay open
        .pingInterval(25, TimeUnit.SECONDS)
        .build()
    private val json = Json { ignoreUnknownKeys = true }

    private val lock = Any()
    private var active = false
    private var sockets = mutableListOf<WebSocket>()
    private var pingTimer: Timer? = null

    /** symbol ("BTCUSDT") → token ("BTC") */
    private var binanceMap: Map<String, String> = emptyMap()

    /** pair ("GT_USDT") → token ("GT") */
    private var gateMap: Map<String, String> = emptyMap()
    private var stockSymbols: List<String> = emptyList()

    private var onCrypto: ((String, Double) -> Unit)? = null
    private var onStock: ((String, Double) -> Unit)? = null

    fun start(
        binanceMap: Map<String, String>,
        gateMap: Map<String, String>,
        stockSymbols: List<String>,
        onCrypto: (String, Double) -> Unit,
        onStock: (String, Double) -> Unit,
    ) {
        stop()
        synchronized(lock) {
            active = true
            this.binanceMap = binanceMap
            this.gateMap = gateMap
            this.stockSymbols = stockSymbols
            this.onCrypto = onCrypto
            this.onStock = onStock
        }
        connectBinance()
        connectGate()
        connectAlpaca()
    }

    fun stop() {
        val open: List<WebSocket>
        synchronized(lock) {
            active = false
            pingTimer?.cancel()
            pingTimer = null
            open = sockets.toList()
            sockets.clear()
        }
        for (socket in open) socket.close(1001, "going away")
    }

    private val isActive: Boolean get() = synchronized(lock) { active }

    // MARK: Plumbing

    private fun open(
        url: String,
        onOpen: ((WebSocket) -> Unit)? = null,
        onMessage: (WebSocket, String) -> Unit,
        reconnect: () -> Unit,
    ) {
        val request = Request.Builder().url(url).build()
        val listener = object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: okhttp3.Response) {
                onOpen?.invoke(webSocket)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                if (!isActive) return
                onMessage(webSocket, text)
            }

            override fun onFailure(
                webSocket: WebSocket,
                t: Throwable,
                response: okhttp3.Response?,
            ) {
                scheduleReconnect(reconnect)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                scheduleReconnect(reconnect)
            }
        }
        synchronized(lock) {
            if (!active) return // stopped mid-reconnect
            sockets.add(client.newWebSocket(request, listener))
        }
    }

    private fun scheduleReconnect(reconnect: () -> Unit) {
        if (!isActive) return
        Timer().schedule(object : TimerTask() {
            override fun run() {
                if (isActive) reconnect()
            }
        }, 5000)
    }

    private fun parse(text: String): JsonObject? = try {
        json.parseToJsonElement(text) as? JsonObject
    } catch (_: Exception) {
        null
    }

    // MARK: Binance — combined miniTicker stream

    private fun connectBinance() {
        val map = synchronized(lock) { binanceMap }
        if (map.isEmpty()) return
        val streams = map.keys
            .map { "${it.lowercase()}@miniTicker" }
            .sorted()
            .joinToString("/")
        open(
            "wss://stream.binance.com:9443/stream?streams=$streams",
            onMessage = { _, text ->
                val root = parse(text) ?: return@open
                val data = root["data"] as? JsonObject ?: return@open
                val symbol = data["s"]?.jsonPrimitive?.contentOrNull ?: return@open
                val token = synchronized(lock) { binanceMap[symbol] } ?: return@open
                val close = data["c"]?.jsonPrimitive?.contentOrNull?.toDoubleOrNull()
                    ?: return@open
                onCrypto?.invoke(token, close)
            },
            reconnect = { connectBinance() },
        )
    }

    // MARK: Gate.io — spot.tickers channel

    private fun connectGate() {
        val map = synchronized(lock) { gateMap }
        if (map.isEmpty()) return
        open(
            "wss://api.gateio.ws/ws/v4/",
            onOpen = { socket ->
                val subscribe = buildJsonObject {
                    put("time", System.currentTimeMillis() / 1000)
                    put("channel", "spot.tickers")
                    put("event", "subscribe")
                    putJsonArray("payload") { map.keys.forEach { add(kotlinx.serialization.json.JsonPrimitive(it)) } }
                }
                socket.send(subscribe.toString())
                // Gate closes idle connections — app-level ping keeps it alive.
                synchronized(lock) {
                    pingTimer?.cancel()
                    pingTimer = Timer().also { timer ->
                        timer.scheduleAtFixedRate(object : TimerTask() {
                            override fun run() {
                                val ping =
                                    "{\"time\":${System.currentTimeMillis() / 1000},\"channel\":\"spot.ping\"}"
                                socket.send(ping)
                            }
                        }, 20_000, 20_000)
                    }
                }
            },
            onMessage = { _, text ->
                val root = parse(text) ?: return@open
                if (root["channel"]?.jsonPrimitive?.contentOrNull != "spot.tickers") return@open
                if (root["event"]?.jsonPrimitive?.contentOrNull != "update") return@open
                val result = root["result"] as? JsonObject ?: return@open
                val pair = result["currency_pair"]?.jsonPrimitive?.contentOrNull ?: return@open
                val token = synchronized(lock) { gateMap[pair] } ?: return@open
                val last = result["last"]?.jsonPrimitive?.contentOrNull?.toDoubleOrNull()
                    ?: return@open
                onCrypto?.invoke(token, last)
            },
            reconnect = { connectGate() },
        )
    }

    // MARK: Alpaca — iex trades

    private fun connectAlpaca() {
        val symbols = synchronized(lock) { stockSymbols }
        if (symbols.isEmpty()) return
        open(
            "wss://stream.data.alpaca.markets/v2/iex",
            onOpen = { socket ->
                socket.send(
                    "{\"action\":\"auth\",\"key\":\"$ALPACA_KEY\",\"secret\":\"$ALPACA_SECRET\"}"
                )
            },
            onMessage = { socket, text ->
                val messages = try {
                    json.parseToJsonElement(text) as? JsonArray
                } catch (_: Exception) {
                    null
                } ?: return@open
                for (element in messages) {
                    val message = element as? JsonObject ?: continue
                    when (message["T"]?.jsonPrimitive?.contentOrNull) {
                        "success" -> {
                            if (message["msg"]?.jsonPrimitive?.contentOrNull == "authenticated") {
                                val subscribe = buildJsonObject {
                                    put("action", "subscribe")
                                    putJsonArray("trades") {
                                        symbols.forEach { add(kotlinx.serialization.json.JsonPrimitive(it)) }
                                    }
                                }
                                socket.send(subscribe.toString())
                            }
                        }
                        "t" -> {
                            val symbol = message["S"]?.jsonPrimitive?.contentOrNull
                            val price = message["p"]?.jsonPrimitive?.doubleOrNull
                            if (symbol != null && price != null) {
                                onStock?.invoke(symbol, price)
                            }
                        }
                    }
                }
            },
            reconnect = { connectAlpaca() },
        )
    }
}
